### PR TITLE
IGNITE-16509 : Calcite engine. Support OTHER data type

### PR DESCRIPTION
**Pull request by Vladsz83**
Table with OTHER (Object) data type can be created by H2-based SQL engine:

`CREATE TABLE t(val OTHER)`

But such a data type is not supported by Calcite-based SQL engine (at least in DDL)

### CODE REVIEW

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeFactory.java [hunk 1]
```diff
@@ -266,22 +266,36 @@ else if (type instanceof UuidType)
...
-    /** @return UUID SQL type. */
-    public RelDataType createUuidType() {
-        return canonize(new UuidType(true));
+    /** @return Custom type by name or storeable class name. {@code Null} if type not found. */
+    public RelDataType createCustomType(String typeName) {
+        return createCustomType(typeName, true);
+    }
+
+    /** @return Nullable custom type by name or storeable class name. {@code Null} if type not found. */
+    public RelDataType createCustomType(String typeName, boolean nullable) {
```
Comment (alex-plekhanov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeFactory.java:286]: I don't like the idea of using the same parameter in different ways (as type name and as class name), perhaps it's better to use some enum. At least we should get rid of literal "UUID" and "OTHER" and use constants.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeFactory.java [hunk 2]
```diff
@@ -266,22 +266,36 @@ else if (type instanceof UuidType)
...
+            return new OtherType(nullable);
+
+        return null;
     }
 
     /** {@inheritDoc} */
     @Override public RelDataType createTypeWithNullability(RelDataType type, boolean nullable) {
-        if (type instanceof UuidType && type.isNullable() != nullable)
-            type = new UuidType(nullable);
+        if (type instanceof IgniteSqlCalciteType && type.isNullable() != nullable)
```
Comment (alex-plekhanov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeFactory.java:297]: `createType` method also should be modified (`UUID` and `Object` classes should be added, and probably `byte[]` for BINARY type). To reproduce the problem without modifying `createType()` method you can add new column to the `JdbcQueryTest#testQueryColumnTypes` test, for example:
```
            "binary_col BINARY, " +
            "uuid_col UUID, " +
            "other_col OTHER, " +
```
And check:
```
            assertEquals(Types.BINARY, md.getColumnType(18));
            assertEquals(byte[].class.getName(), md.getColumnClassName(18));
            assertEquals(Types.OTHER, md.getColumnType(19));
            assertEquals(UUID.class.getName(), md.getColumnClassName(19));
            assertEquals(Types.OTHER, md.getColumnType(20));
            assertEquals(Object.class.getName(), md.getColumnClassName(20));
```
Currently, all these columns return `Types.OTHER`  as column type and `Object[]` as column class name

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/jdbc/JdbcQueryTest.java [hunk 3]
```diff
@@ -87,6 +98,62 @@
...
+     * @throws SQLException If failed.
+     */
+    @Test
+    public void testOtherType() throws Exception {
+        List<Object> values = new ArrayList<>();
+
+        values.add("str");
+        values.add(11);
+        values.add(101.1);
+        values.add(202.2f);
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/jdbc/JdbcQueryTest.java:111]: Yep. Is double now

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/jdbc/JdbcQueryTest.java [hunk 4]
```diff
@@ -87,6 +98,62 @@
...
+
+                assertEquals(1, ps.executeUpdate());
+            }
+        }
+
+        try (ResultSet rs = stmt.executeQuery("select * from tbl order by id")) {
+            for (Object obj : values) {
+                assertTrue(rs.next());
+
+                if (obj != null && obj.getClass().isArray())
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/jdbc/JdbcQueryTest.java:149]: Fixed

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/jdbc/JdbcQueryTest.java [hunk 5]
```diff
@@ -266,4 +333,38 @@ public void testQueryColumnTypes() throws Exception {
...
+        }
+
+        /** */
+        @Override public boolean equals(Object o) {
+            if (this == o)
+                return true;
+            if (o == null || getClass() != o.getClass())
+                return false;
+            ObjectToStore store = (ObjectToStore)o;
+            return id == store.id && Double.compare(store.val, val) == 0 && Objects.equals(name, store.name);
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/jdbc/JdbcQueryTest.java:362]: Fixed

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/DataTypesTest.java [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+import org.apache.ignite.configuration.CacheConfiguration;
+import org.apache.ignite.internal.processors.query.IgniteSQLException;
+import org.apache.ignite.internal.processors.query.calcite.integration.AbstractBasicIntegrationTest;
+import org.apache.ignite.internal.util.typedef.F;
+import org.junit.Test;
+
+/**
+ * Test SQL data types.
+ */
+public class DataTypesTest extends AbstractBasicIntegrationTest {
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/DataTypesTest.java:37]: Fixed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteSqlCalciteType.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+package org.apache.ignite.internal.processors.query.calcite.type;
+
+import java.lang.reflect.Type;
+import org.apache.calcite.rel.type.RelDataTypeFamily;
+import org.apache.calcite.rel.type.RelDataTypeImpl;
+import org.apache.calcite.sql.type.SqlTypeFamily;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+/** Custom type base. */
+public abstract class IgniteSqlCalciteType extends RelDataTypeImpl {
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteSqlCalciteType.java:27]: Fixed

File: modules/calcite/src/main/codegen/includes/parserImpls.ftl [hunk 8]
```diff
@@ -99,6 +99,8 @@ SqlDataTypeSpec DataTypeEx() :
         dt = IntervalType()
     |
         dt = UuidType()
+    |
+        dt = OtherType()
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/calcite/src/main/codegen/includes/parserImpls.ftl:103]: Hm. Fixed.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeFactory.java (hunk 1 above)
Comment (Vladsz83) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeFactory.java:286]: Changed to Type

File: modules/calcite/src/main/codegen/config.fmpp [hunk 9]
```diff
@@ -68,6 +68,7 @@ data: {
       "ASYNC"
       "QUERY"
       "UUID"
+      "OTHER"
```
Comment (alex-plekhanov) [inline modules/calcite/src/main/codegen/config.fmpp:71]: These changes redundant too

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeFactory.java [hunk 10]
```diff
@@ -266,22 +266,36 @@ else if (type instanceof UuidType)
...
+    public RelDataType createCustomType(Type type) {
+        return createCustomType(type, true);
+    }
+
+    /** @return Nullable custom type by storage type. {@code Null} if custom type not found. */
+    public RelDataType createCustomType(Type type, boolean nullable) {
+        if (UUID.class == type)
+            return new UuidType(nullable);
+        else if (Object.class == type)
+            return new OtherType(nullable);
```
Comment (alex-plekhanov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeFactory.java:290]: `return canonize(new ...)`

File: modules/calcite/src/main/codegen/config.fmpp (hunk 9 above)
Comment (Vladsz83) [reply modules/calcite/src/main/codegen/config.fmpp]: Fixed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeFactory.java (hunk 10 above)
Comment (Vladsz83) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeFactory.java]: Fixed

**Review by alex-plekhanov — APPROVED** [review]

**Pull request closed without merging**